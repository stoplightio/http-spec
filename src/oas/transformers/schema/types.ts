import type { JSONSchema7 } from 'json-schema';
import type { SchemaObject } from 'openapi3-ts/src/model/OpenApi';
import type { Schema } from 'swagger-schema-official';

export type OASSchemaObject = SchemaObject | Schema;

export type Converter = (schema: OASSchemaObject | JSONSchema7) => void;
