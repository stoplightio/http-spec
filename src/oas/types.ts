import { DeepPartial, IHttpOperation, IHttpService } from '@stoplight/types';
import { OpenAPIObject } from 'openapi3-ts';
import { Spec } from 'swagger-schema-official';

export const enum OasVersion {
  OAS2 = 2,
  OAS3 = 3,
}

/**
 * Service
 */

export interface ITransformServiceOpts<T> {
  document: DeepPartial<T>;
}

export type HttpServiceTransformer<T> = (opts: T) => IHttpService;

export type Oas2TransformServiceOpts = ITransformServiceOpts<Spec>;
export type Oas3TransformServiceOpts = ITransformServiceOpts<OpenAPIObject>;
export type Oas2HttpServiceTransformer = HttpServiceTransformer<Oas2TransformServiceOpts>;
export type Oas3HttpServiceTransformer = HttpServiceTransformer<Oas3TransformServiceOpts>;

/**
 * Operation
 */

export interface ITransformOperationOpts<T> {
  document: DeepPartial<T>;
  path: string;
  method: string;
}

export type HttpOperationTransformer<T> = (opts: T) => IHttpOperation;

export type Oas2TransformOperationOpts = ITransformOperationOpts<Spec>;
export type Oas3TransformOperationOpts = ITransformOperationOpts<OpenAPIObject>;
export type Oas2HttpOperationTransformer = HttpOperationTransformer<Oas2TransformOperationOpts>;
export type Oas3HttpOperationTransformer = HttpOperationTransformer<Oas3TransformOperationOpts>;
