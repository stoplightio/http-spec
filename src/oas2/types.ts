import type { DeepPartial } from '@stoplight/types';
import type { Spec } from 'swagger-schema-official';

import { TranslateFunction } from '../types';

export type Oas2TranslateFunction<P extends unknown[], R extends unknown = unknown> = TranslateFunction<
  DeepPartial<Spec>,
  P,
  R
>;
