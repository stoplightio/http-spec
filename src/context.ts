import { hasRef, isLocalRef } from '@stoplight/json';
// @ts-ignore
import * as fnv from 'fnv-plus';

import type {
  AvailableContext,
  Fragment,
  IdGenerator,
  RefResolver,
  TransformerContext,
  TranslateFunction,
} from './types';

export const DEFAULT_ID_GENERATOR: IdGenerator = input => fnv.fast1a32hex(input);

export function createContext<T extends Record<string, unknown>>(
  document: T,
  resolveRef: RefResolver<T>,
  generateId: IdGenerator,
): TransformerContext<T> {
  let context: AvailableContext = 'service';
  return {
    document,
    get context() {
      return context;
    },
    set context(value) {
      context = value;
      if (value !== 'operation') {
        this.parentId = this.ids[value];
      }
    },
    maybeResolveLocalRef(target: unknown) {
      if (hasRef(target) && isLocalRef(target.$ref)) {
        try {
          return resolveRef.call(this, target);
        } catch {
          return target;
        }
      }

      return target;
    },
    generateId(value) {
      this.parentId = generateId(value);
      return this.parentId;
    },
    ids: {
      service: '',
      path: '',
      operation: '',
    },
    parentId: '',
  };
}

export function withContext<F extends TranslateFunction<any, any[]> = TranslateFunction<Fragment, unknown[]>>(
  fn: F,
): F {
  return <F>function (...args: Parameters<F>) {
    const { context, parentId } = this;
    const result = fn.apply(this, args);
    this.context = context;
    this.parentId = parentId;
    return result;
  };
}
