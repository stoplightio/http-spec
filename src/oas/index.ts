export * from './operation';
export * from './types';
