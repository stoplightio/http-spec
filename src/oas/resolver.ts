import type { JsonPath } from '@stoplight/types';

import type { AvailableContext, RefResolver } from '../types';
import { resolveInlineRef } from './resolveInlineRef';

function inferContext(path: JsonPath): AvailableContext {
  if (path.length < 2 || path[0] !== 'paths') return 'service';
  if (path.length === 2 || path[3] === 'parameters' || path[3] === 'servers') return 'path';
  return 'operation';
}

const SHARED_COMPONENTS_KEYS = new WeakMap();

export function getSharedKey(value: object) {
  return SHARED_COMPONENTS_KEYS.get(value);
}

export const resolveRef: RefResolver = function (target) {
  const location: string[] = [];
  const resolved = resolveInlineRef(this.document, target.$ref, location);

  const context = inferContext(location);
  if (context !== null && this.context !== context) {
    this.context = context;
  }

  if (typeof resolved === 'object' && resolved !== null && context === 'service') {
    SHARED_COMPONENTS_KEYS.set(resolved, location[0] === 'components' ? location[2] : location[1]);
  }

  return resolved;
};
