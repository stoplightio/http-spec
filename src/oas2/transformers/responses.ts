import { isPlainObject } from '@stoplight/json';
import type { IHttpOperationResponse, Optional } from '@stoplight/types';
import { DeepPartial } from '@stoplight/types';
import { Operation } from 'swagger-schema-official';

import { isNonNullable } from '../../guards';
import { translateSchemaObject } from '../../oas/transformers/schema';
import { entries } from '../../utils';
import { getExamplesFromSchema, getProduces } from '../accessors';
import { isResponseObject } from '../guards';
import { Oas2TranslateFunction } from '../types';
import { translateToHeaderParams } from './params';

const translateToResponse: Oas2TranslateFunction<
  [produces: string[], statusCode: string, response: unknown],
  Optional<IHttpOperationResponse>
> = function (produces, statusCode, response) {
  const resolvedResponse = this.maybeResolveLocalRef(response);
  if (!isResponseObject(resolvedResponse)) return;

  this.state.enter('responses', statusCode);

  const headers = translateToHeaderParams.call(this, resolvedResponse.headers);
  const objectifiedExamples = entries(
    resolvedResponse.examples || (resolvedResponse.schema ? getExamplesFromSchema(resolvedResponse.schema) : void 0),
  ).map(([key, value]) => ({ id: this.generateId('example'), key, value }));

  const contents = produces
    .map(produceElement => ({
      id: this.generateId('security'),
      mediaType: produceElement,
      schema: isPlainObject(resolvedResponse.schema)
        ? translateSchemaObject.call(this, resolvedResponse.schema)
        : void 0,
      examples: objectifiedExamples.filter(example => example.key === produceElement),
    }))
    .filter(({ schema, examples }) => !!schema || examples.length > 0);

  const translatedResponses = {
    id: this.generateId('security'),
    code: statusCode,
    description: resolvedResponse.description,
    headers,
    contents,
  };

  const foreignExamples = objectifiedExamples.filter(example => !produces.includes(example.key));
  if (foreignExamples.length > 0) {
    if (translatedResponses.contents.length === 0)
      translatedResponses.contents[0] = {
        id: this.generateId('param'),
        mediaType: '',
        schema: {},
        examples: [],
      };

    translatedResponses.contents[0].examples!.push(...foreignExamples);
  }

  return translatedResponses;
};

export const translateToResponses: Oas2TranslateFunction<
  [operation: DeepPartial<Operation>],
  IHttpOperationResponse[]
> = function (operation) {
  const produces = getProduces(this.document, operation);
  return entries(operation.responses)
    .map(([statusCode, response]) => translateToResponse.call(this, produces, statusCode, response))
    .filter(isNonNullable);
};
