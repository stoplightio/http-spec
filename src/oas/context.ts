// @ts-expect-error: no types
import * as fnv from 'fnv-plus';

import { createContext as _createContext } from '../context';
import type { Fragment, IdGenerator } from '../types';
import { resolveRef } from './resolver';

const HASH_32_KINDS = ['http_operation', 'schema'];

function getKind(input: string) {
  const i = input.indexOf('-');
  return input.slice(0, i);
}

const DEFAULT_ID_GENERATOR: IdGenerator = input => {
  const kind = getKind(input);
  return HASH_32_KINDS.some(k => kind === k) ? fnv.fast1a32hex(input) : fnv.fast1a52hex(input);
};

export function createContext<T extends Fragment>(document: T) {
  return _createContext<T>(document, resolveRef, DEFAULT_ID_GENERATOR);
}
