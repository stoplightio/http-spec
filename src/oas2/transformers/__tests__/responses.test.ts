import { DeepPartial } from '@stoplight/types';
import { Operation, Schema, Spec } from 'swagger-schema-official';

import { createContext, DEFAULT_ID_GENERATOR } from '../../../context';
import { resolveRef } from '../../../oas/resolver';
import { translateToResponses as _translateToResponses } from '../responses';

const translateToResponses = (document: DeepPartial<Spec>, responses: DeepPartial<Operation['responses']>) =>
  _translateToResponses.call(createContext(document, resolveRef, DEFAULT_ID_GENERATOR), { responses });

describe('responses', () => {
  const produces = ['application/json', 'application/xml'];

  it('should translate to multiple responses', () => {
    const responses = translateToResponses(
      { produces },
      {
        r1: {
          description: 'd1',
          examples: {
            'application/json': {},
          },
          headers: {},
          schema: {},
        },
        r2: {
          description: 'd2',
          examples: {
            'application/xml': {},
          },
          headers: {},
          schema: {},
        },
      },
    );

    expect(responses).toHaveLength(2);
    expect(responses[0]).toMatchSnapshot({
      id: expect.any(String),
      contents: [
        {
          id: expect.any(String),
          schema: {
            'x-stoplight-id': expect.any(String),
          },
          examples: [
            {
              id: expect.any(String),
            },
          ],
        },
        {
          id: expect.any(String),
          schema: {
            'x-stoplight-id': expect.any(String),
          },
          examples: [],
        },
      ],
    });
    expect(responses[1]).toMatchSnapshot({
      id: expect.any(String),
      contents: [
        {
          id: expect.any(String),
          schema: {
            'x-stoplight-id': expect.any(String),
          },
        },
        {
          id: expect.any(String),
          schema: {
            'x-stoplight-id': expect.any(String),
          },
          examples: [
            {
              id: expect.any(String),
            },
          ],
        },
      ],
    });
  });

  it('should translate to response w/o headers', () => {
    const responses = translateToResponses(
      { produces },
      {
        r1: {
          description: 'd1',
          examples: {
            'application/xml': {},
          },
          schema: {},
        },
      },
    );

    expect(responses).toHaveLength(1);
    expect(responses[0]).toMatchSnapshot({
      id: expect.any(String),
      contents: [
        {
          id: expect.any(String),
          schema: {
            'x-stoplight-id': expect.any(String),
          },
        },
        {
          id: expect.any(String),
          schema: {
            'x-stoplight-id': expect.any(String),
          },
          examples: [
            {
              id: expect.any(String),
            },
          ],
        },
      ],
    });
  });

  it('should translate to response w/o examples', () => {
    const responses = translateToResponses(
      { produces },
      {
        r1: {
          description: 'd1',
          schema: {},
        },
      },
    );

    expect(responses).toHaveLength(1);
    expect(responses[0]).toMatchSnapshot({
      id: expect.any(String),
      contents: [
        {
          id: expect.any(String),
          schema: {
            'x-stoplight-id': expect.any(String),
          },
        },
        {
          id: expect.any(String),
          schema: {
            'x-stoplight-id': expect.any(String),
          },
        },
      ],
    });
  });

  describe('should keep foreign examples', () => {
    it('aggregating them to the first example', () => {
      const responses = translateToResponses(
        { produces },
        {
          r1: {
            description: 'd1',
            examples: {
              'application/i-have-no-clue': {},
              'application/json': {},
            },
            headers: {},
            schema: {},
          },
        },
      );

      expect(responses[0].contents).toBeDefined();
      expect(responses[0].contents![0]).toHaveProperty('mediaType', 'application/json');

      expect(responses[0].contents![0].examples).toBeDefined();
      expect(responses[0].contents![0].examples).toHaveLength(2);
    });
  });

  describe('schema examples', () => {
    describe('given a response with a schema with an example', () => {
      it('should translate to response with examples', () => {
        const responses = translateToResponses(
          { produces },
          {
            r1: {
              description: 'd1',
              headers: {},
              schema: {
                example: {
                  name: 'value',
                },
              },
            },
          },
        );
        expect(responses[0].contents![0]).toHaveProperty('examples', [
          { id: expect.any(String), key: 'default', value: { name: 'value' } },
        ]);
      });
    });

    describe('given multiple schema example properties', () => {
      it('should translate all examples', () => {
        const responses = translateToResponses(
          { produces },
          {
            r1: {
              description: 'd1',
              headers: {},
              schema: {
                example: {
                  name: 'example value',
                },
                ['x-examples']: {
                  'application/json': {
                    name: 'examples value',
                  },
                },
              } as Schema,
            },
          },
        );
        expect(responses[0].contents![0]).toHaveProperty('examples', [
          { id: expect.any(String), key: 'application/json', value: { name: 'examples value' } },
          { id: expect.any(String), key: 'default', value: { name: 'example value' } },
        ]);
      });
    });

    describe('given response with examples in root and schema objects', () => {
      it('root examples should take precedence over schema examples', () => {
        const responses = translateToResponses(
          { produces },
          {
            r1: {
              description: 'd1',
              headers: {},
              examples: {
                'application/i-have-no-clue': {},
                'application/json': {},
              },
              schema: {
                example: {
                  name: 'value',
                },
              },
            },
          },
        );
        expect(responses[0].contents![0].examples).toHaveLength(2);
        expect(responses[0].contents![0].examples).toEqual([
          { id: expect.any(String), key: 'application/json', value: {} },
          { id: expect.any(String), key: 'application/i-have-no-clue', value: {} },
        ]);
      });
    });
  });
});
