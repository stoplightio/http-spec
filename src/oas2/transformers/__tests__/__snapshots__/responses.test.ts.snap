// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`responses should translate to multiple responses 1`] = `
Object {
  "code": "r1",
  "contents": Array [
    Object {
      "examples": Array [
        Object {
          "id": "5b16088e",
          "key": "application/json",
          "value": Object {},
        },
      ],
      "id": Any<String>,
      "mediaType": "application/json",
      "schema": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "x-stoplight-id": Any<String>,
      },
    },
    Object {
      "examples": Array [],
      "id": Any<String>,
      "mediaType": "application/xml",
      "schema": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "x-stoplight-id": Any<String>,
      },
    },
  ],
  "description": "d1",
  "headers": Array [],
  "id": Any<String>,
}
`;

exports[`responses should translate to multiple responses 2`] = `
Object {
  "code": "r2",
  "contents": Array [
    Object {
      "examples": Array [],
      "id": Any<String>,
      "mediaType": "application/json",
      "schema": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "x-stoplight-id": Any<String>,
      },
    },
    Object {
      "examples": Array [
        Object {
          "id": Any<String>,
          "key": "application/xml",
          "value": Object {},
        },
      ],
      "id": Any<String>,
      "mediaType": "application/xml",
      "schema": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "x-stoplight-id": Any<String>,
      },
    },
  ],
  "description": "d2",
  "headers": Array [],
  "id": Any<String>,
}
`;

exports[`responses should translate to response w/o examples 1`] = `
Object {
  "code": "r1",
  "contents": Array [
    Object {
      "examples": Array [],
      "id": Any<String>,
      "mediaType": "application/json",
      "schema": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "x-stoplight-id": Any<String>,
      },
    },
    Object {
      "examples": Array [],
      "id": Any<String>,
      "mediaType": "application/xml",
      "schema": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "x-stoplight-id": Any<String>,
      },
    },
  ],
  "description": "d1",
  "headers": Array [],
  "id": Any<String>,
}
`;

exports[`responses should translate to response w/o headers 1`] = `
Object {
  "code": "r1",
  "contents": Array [
    Object {
      "examples": Array [],
      "id": Any<String>,
      "mediaType": "application/json",
      "schema": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "x-stoplight-id": Any<String>,
      },
    },
    Object {
      "examples": Array [
        Object {
          "id": Any<String>,
          "key": "application/xml",
          "value": Object {},
        },
      ],
      "id": Any<String>,
      "mediaType": "application/xml",
      "schema": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "x-stoplight-id": Any<String>,
      },
    },
  ],
  "description": "d1",
  "headers": Array [],
  "id": Any<String>,
}
`;
